### PR TITLE
Tcp echo server

### PR DESCRIPTION
**Pull request by savosaicic**
Implement the tcp echo server

### CODE REVIEW

File: srcs/webserv.cpp [hunk 1]
```diff
@@ -6,27 +6,62 @@
...
+  {
+    if (stream.read() == 0)
+    {
+      std::cout << "disconnected client" << std::endl;
+      listener.dropClient(stream);
+    }
+    else
+    {
+      std::cout << "message receive: " << stream.getInMessage() << std::endl;
+      // listener.unRegisterClientEvent(stream, EPOLLIN);
```
Comment (theovgl) [inline srcs/webserv.cpp:33]: ?????

File: includes/TCPListener.hpp [hunk 2]
```diff
@@ -21,17 +21,26 @@ class TCPListener
 {
     int _epfd;
     std::vector<std::pair<int, const char*> > _fd_port_list;
+    std::vector<TCPStream>  _connected_client;
```
Comment (vfurmane) [inline includes/TCPListener.hpp]: :smiling_imp: 
```suggestion
    std::vector<TCPStream>  _connected_clients;
```

File: srcs/IPAdress.cpp [hunk 3]
```diff
@@ -28,6 +30,18 @@ IPAddress::IPAddress(const IPAddress& src) : _addr(src._addr),
 {
 }
 
+IPAddress::IPAddress(int fd)
+{
+  memset(&_addr, 0, sizeof(_addr));
+  memset(&_addrlen, 0, sizeof(_addrlen));
+  if (getpeername(fd, &_addr, &_addrlen) != 0)
```
Comment (lefendie) [inline srcs/IPAdress.cpp]: getpeername pas dans le sujet, faut demander avant de l'utiliser je pense

File: includes/TCPListener.hpp (hunk 2 above)
Comment (vfurmane) [inline includes/TCPListener.hpp]: A manager class could be more convenient.

If you had to send the whole connected clients list, including their methods, you would need to send the whole TCPListener. Whereas a manager allows you to use this list regardless from the server it belongs to.

File: includes/TCPStream.hpp [hunk 4]
```diff
@@ -16,33 +16,46 @@
...
+		uint32_t triggeredEvents() const;
+    uint32_t getRegisteredEvents() const;
+		const std::string& getInMessage() const;
+		const std::string& getOutMessage() const;
+
+    void  clearInMessage();
+    void  clearOutMessage();
+
+    int read();
+    int write(const std::string& str);
```
Comment (vfurmane) [inline includes/TCPStream.hpp:56]: What about using the `<<` and `>>` operators :)

**Review by ArnaudBD — APPROVED** [review]

File: srcs/TCPListener.cpp [hunk 5]
```diff
@@ -89,39 +109,108 @@ bool TCPListener::isListener(int fd)
...
+}
+
+std::vector<TCPStream>::iterator  TCPListener::_findClient(int fd)
+{
+  std::vector<TCPStream>::iterator it(_connected_client.begin());
+  for (; it != _connected_client.end(); ++it)
+  {
+    if (it->getFd() == fd)
+      break;
+  }
```
Comment (vfurmane) [inline srcs/TCPListener.cpp]: Use `std::find` ?

**Review by vfurmane — APPROVED** [review]

**Review by lefendie — APPROVED** [review]

**Review by lefendie — APPROVED** [review]

**Pull request merged**